### PR TITLE
Support using notes via a commit rather than a ref

### PR DESCRIPTION
**Pull request by richardipsum**
Hi!

This is a work in progress implementation to solve issue #3782 
each new function will be added in a separate commit.

Finally existing notes functions will be reimplemented in terms of the new functions.

I'm opening this now in case anyone has any suggestions in the early stages
that will allow me to avoid rework later on.

Thanks,
Richard Ipsum

### CODE REVIEW

**Comment by richardipsum** [comment]
@joshtriplett @ethomson I am not certain how to go about the last stage of this feature. It's not clear how to have, for example, `git_note_create` wrap `git_note_create_from_commit` since `git_note_create` returns the oid of the note blob, which is not returned by `git_note_create_from_commit`.

I really have no idea what the best option would be here, other than just leaving the existing functions as they are rather than trying to make them wrap the `from_commit` functions.

Ideas and suggestions would be really helpful at this point!

**Comment by joshtriplett** [comment]
@richardipsum While it seems somewhat unnecessary for most applications, returning the new blob OID as an additional out parameter seems fine.  (A caller could pass NULL if they don't care about the blob OID.)

**Comment by richardipsum** [comment]
@joshtriplett That would work, okay I'll do that then, thanks :)

**Comment by richardipsum** [comment]
@ethomson I'm not sure what the test failure here is really about, but it seems unrelated to my change?

**Comment by ethomson** [comment]
@joshtriplett can you provide some feedback here?  Does this address your problem(s)?  Are you reasonably happy with the API?  Are there things this doesn't do that you wish it did?  (Or things that it does do that you think are unnecessary?)

**Comment by ethomson** [comment]
Also, I'm not super familiar with notes.  It's literally a feature that I've never used... 

Here's my big question that I don't understand yet:  why is our existing API dealing with refs?  It looks like it should be dealing with commits in the first place, doesn't it?  Skimming the `git notes` documentation, it looks like `git` itself deals with commits...?

Is our existing API useful at all?  Ignoring API compatibility for a moment, should we replace what we have with this?

**Comment by joshtriplett** [comment]
@ethomson Yes, this does exactly what I need it to do.

I'd assume the existing API deals with refs because by default, `git notes` uses a ref under refs/notes/commits to accumulate the notes. So, it makes sense to retain APIs that make it easy to say "add a note to the default notes ref". The API I needed was for the case where you want to store and share notes a different way, unlike the `git notes` tool.

**Comment by pks-t** [comment]
@ethomson You got that slightly wrong. Notes do not even work on commits, but instead on arbitrary objects. Will comment on the usefulness of our API in a few minutes, as soon as I've reviewed this PR

File: include/git2/notes.h [hunk 1]
```diff
@@ -153,6 +153,33 @@ GIT_EXTERN(int) git_note_create(
...
+GIT_EXTERN(int) git_note_commit_create(
+	git_commit **notes_commit_out,
+	git_oid *notes_blob_out,
+	git_repository *repo,
+	git_commit *notes_commit,
+	const git_signature *author,
+	const git_signature *committer,
+	const git_oid *oid,
+	const char *note,
+	int allow_note_overwrite);
```
Comment (pks-t) [inline include/git2/notes.h]: I'm currently trying to understand this function. As I understand, a note is created for an object (e.g. either a blob, tree, commit or tag) inside of a specific reference. At least, this is how I understand `git notes add`.

So this function here removes the note reference parameter from `git_note_create` which hints where the note should be put in, and instead uses an existing notes commit as its parent. And subsequently, we do not store the resulting notes commit in any reference, instead it's created as a dangling commit.

So this buys us the ability to easily create a complete notes tree without having to write and re-read the notes reference, correct? And in the end, the developer is free to create the notes reference for himself.

File: include/git2/notes.h [hunk 2]
```diff
@@ -153,6 +153,33 @@ GIT_EXTERN(int) git_note_create(
...
+ *					NULL in case of error
+ * @param notes_blob_out a point to the id of a note blob (optional)
+ * @param repo repository where the note will live
+ * @param notes_commit a pointer to the notes commit object
+ *					or NULL if this is to be the first notes object
+ * @param author signature of the notes commit author
+ * @param committer signature of the notes commit committer
+ * @param oid OID of the git object to decorate
+ * @param note Content of the note to add for object oid
+ * @param force Overwrite existing note
```
Comment (pks-t) [inline include/git2/notes.h]: How can this function possibly overwrite an existing note if no notes reference is created?

File: include/git2/notes.h [hunk 3]
```diff
@@ -153,6 +153,33 @@ GIT_EXTERN(int) git_note_create(
...
 
+/**
+ * Add a note for an object from a commit
+ *
+ * @param notes_commit_out pointer to store the commit (optional);
+ *					NULL in case of error
+ * @param notes_blob_out a point to the id of a note blob (optional)
+ * @param repo repository where the note will live
+ * @param notes_commit a pointer to the notes commit object
+ *					or NULL if this is to be the first notes object
```
Comment (pks-t) [inline include/git2/notes.h]: I think naming this `parent` would be more insightful, wouldn't it?

```
@param parent Pointer to parent note or NULL if this shall start a new notes tree
```

File: include/git2/notes.h [hunk 4]
```diff
@@ -153,6 +153,33 @@ GIT_EXTERN(int) git_note_create(
 	const char *note,
 	int force);
 
+/**
+ * Add a note for an object from a commit
```
Comment (pks-t) [inline include/git2/notes.h]: It would be insightful if this would note that no actual notes reference is written.

Comment (pks-t) [inline include/git2/notes.h]: Or maybe even better: adopt an optional notes reference which is to be written. This would ease finalizing the notes tree.

File: src/notes.c [hunk 5]
```diff
@@ -491,6 +497,50 @@ int git_note_create(
...
+			note, tree, target, &notes_commit, allow_note_overwrite);
+
+	if (error < 0)
+		goto cleanup;
+
+	if (notes_commit_out != NULL) {
+		error = git_commit_lookup(notes_commit_out, repo, &note_commit_oid);
+		if (error < 0)
+			goto cleanup;
+	}
```
Comment (pks-t) [inline src/notes.c]: So we actually do a separate lookup on this commit. I'd instead propose simply giving out the OID of the commit, as with the blob ID, and have the caller handle the lookup in case it is required.

File: include/git2/notes.h (hunk 1 above)
Comment (joshtriplett) [inline include/git2/notes.h]: @pks-t Yes, that's the exact use case.

File: include/git2/notes.h (hunk 4 above)
Comment (joshtriplett) [inline include/git2/notes.h]: @pks-t Adding an *optional* notes reference does seem convenient, but on the other hand, then you'd want the ability to use the current value of that reference as the parent, and that starts to make this API look complicated.

File: src/notes.c (hunk 5 above)
Comment (joshtriplett) [inline src/notes.c]: @pks-t Agreed.

**Comment by richardipsum** [comment]
@pks-t  Thanks for your review, I need to think carefully about what's been said here, I will do so and reply once that's done.

File: include/git2/notes.h (hunk 4 above)
Comment (pks-t) [inline include/git2/notes.h]: @joshtriplett I think I agree by now that the signature is fine as-is.

File: include/git2/notes.h (hunk 3 above)
Comment (richardipsum) [inline include/git2/notes.h]: yes

File: include/git2/notes.h (hunk 2 above)
Comment (richardipsum) [inline include/git2/notes.h]: I believe we can overwrite the notes object itself.

    	if ((error = manipulate_note_in_tree_r(
		&tree, repo, commit_tree, &oid, target, 0,
		allow_note_overwrite ? insert_note_in_tree_enotfound_cb : insert_note_in_tree_eexists_cb,
		insert_note_in_tree_enotfound_cb)) < 0)
		goto cleanup;

File: include/git2/notes.h (hunk 1 above)
Comment (richardipsum) [inline include/git2/notes.h]: indeed, exactly

File: src/notes.c (hunk 5 above)
Comment (richardipsum) [inline src/notes.c]: returning the oid would also make this more consistent since `git_commit_create` takes an array of `git_commit` objects and returns an `oid`

**Comment by richardipsum** [comment]
@pks-t I agree with all your points and will make the necessary changes.

**Comment by richardipsum** [comment]
@pks-t 

> I'm a bit on the edge though regarding adding a second set of functions that handle note commits instead of references. It simply feels wrong and is apt to confuse users. The current API is really restricted though in its use and seems a tad mis-fitted for more complex handling of notes. So I tend to think that adjusting the current API to work on commits instead of references would be preferable. Creating the actual notes ref should then be handled via our refs API.

I'm not sure what you're suggesting here, the current PR provides an API that allows more complex handling of notes while still keep backwards compatibility, are you saying you're considering replacing the existing set of functions with these? If so perhaps this is something that could be done in a separate PR, since I imagine it's not possible to merge such an API breaking change without releasing a new major version.

For now I will just make the few small changes you've suggested and not worry about removing the existing notes functions.

**Comment by richardipsum** [comment]
@pks-t I've implemented the changes suggested!

File: tests/notes/notes.c [hunk 6]
```diff
@@ -370,6 +370,46 @@ static char *messages[] = {
 
 #define MESSAGES_COUNT (sizeof(messages)/sizeof(messages[0])) - 1
 
+/* Test that we can read a note */
```
Comment (pks-t) [inline tests/notes/notes.c]: No need for the comment here.

File: tests/notes/notes.c [hunk 7]
```diff
@@ -370,6 +370,46 @@ static char *messages[] = {
...
+
+	cl_assert(notes_commit);
+
+	cl_git_pass(git_note_commit_read(&note, _repo, notes_commit, &oid));
+
+	cl_assert_equal_s(git_note_message(note), "I decorate 4a20\n");
+
+	git_commit_free(notes_commit);
+	git_note_free(note);
+}
```
Comment (pks-t) [inline tests/notes/notes.c]: I'd recommend to also test some edge-cases here. At least a test for a note that cannot be read would be nice to have.

File: include/git2/notes.h [hunk 8]
```diff
@@ -223,6 +223,27 @@ GIT_EXTERN(int) git_note_remove(
 	const git_oid *oid);
 
 /**
+ * Remove the note for an object
+ *
+ * @param notes_commit_out pointer to store the commit (optional);
+ *					NULL in case of error
```
Comment (pks-t) [inline include/git2/notes.h]: Which commit is this exactly? Is this any different from `notes_commit`? I think this concept should be explained a bit better.

File: src/notes.c [hunk 9]
```diff
@@ -587,6 +595,41 @@ int git_note_remove(git_repository *repo, const char *notes_ref_in,
...
+
+	error = note_remove(&note_oid,
+		repo, author, committer, NULL, tree, target, &notes_commit);
+
+	if (error < 0)
+		goto cleanup;
+
+	if (notes_commit_out != NULL) {
+		git_oid_cpy(notes_commit_out, &note_oid);
+	}
```
Comment (pks-t) [inline src/notes.c]: No braces needed.

File: src/notes.c [hunk 10]
```diff
@@ -587,6 +595,41 @@ int git_note_remove(git_repository *repo, const char *notes_ref_in,
...
+		const git_signature *committer,
+		const git_oid *oid)
+{
+	int error;
+	char *target = NULL;
+	git_tree *tree = NULL;
+	git_oid note_oid;
+
+	target = git_oid_allocfmt(oid);
+	GITERR_CHECK_ALLOC(target);
```
Comment (pks-t) [inline src/notes.c]: No need for dynamic allocation here.

**Comment by richardipsum** [comment]
@pks-t Thanks for the review! I should be able to find time to make the requested changes this weekend, thanks!

File: include/git2/notes.h (hunk 8 above)
Comment (richardipsum) [reply include/git2/notes.h]: Yes it should be, when a note is removed we make a new tree with all notes sans the note to be removed, so a new commit is created pointing to this new tree. Even if we remove the last note, we still have a pointer (the commit) to an empty tree object.

File: src/notes.c [hunk 11]
```diff
@@ -463,6 +463,30 @@ int git_note_read(git_note **out, git_repository *repo,
...
+	git_repository *repo,
+	git_commit *notes_commit,
+	const git_oid *oid)
+{
+	int error;
+	char *target = NULL;
+	git_tree *tree = NULL;
+
+	target = git_oid_allocfmt(oid);
+	GITERR_CHECK_ALLOC(target);
```
[the remark this answers is not in the document]
Comment (richardipsum) [reply src/notes.c]: I must have missed it, will switch to static buffer, I also notice that in `git_note_read` `target` is allocated then not used, so will fix that also.

File: src/notes.c [hunk 12]
```diff
@@ -491,6 +497,48 @@ int git_note_create(
...
+	const git_oid *oid,
+	const char *note,
+	int allow_note_overwrite)
+{
+	int error;
+	git_tree *tree = NULL;
+	char *target = NULL;
+	git_oid note_commit_oid, note_blob_oid;
+
+	target = git_oid_allocfmt(oid);
```
[the remark this answers is not in the document]
Comment (richardipsum) [reply src/notes.c]: Not sure I follow your first point about accepting an OID. Do you mean why does `note_write` take target as `char *` rather than `git_oid *` ?

For now I'll just switch to static buffer, and `git_oid_tostr`

**Comment by richardipsum** [comment]
@pks-t I've finally completed the requested changes, I'm sorry it took me such a long time to complete them. :disappointed:

**Comment by pks-t** [comment]
No need to be disappointed. To be honest, I'm so swamped in work right now so that I'm happy whenever it's not one of us being responsible for delayed responses :P Even though your comment is already two weeks old as of now...

File: src/notes.c [hunk 13]
```diff
@@ -497,6 +497,44 @@ int git_note_create(
...
+	git_tree *tree = NULL;
+	git_oid note_commit_oid, note_blob_oid;
+	char target[GIT_OID_HEXSZ + 1];
+
+	git_oid_tostr(target, sizeof(target), oid);
+
+	if (parent != NULL && (error = git_commit_tree(&tree, parent)) < 0)
+		goto cleanup;
+
+	error = note_write(&note_commit_oid, &note_blob_oid, repo, author, committer, NULL,
```
Comment (pks-t) [inline src/notes.c]: Wait a second, this is strange. You're passing in `git_oid **`'s here, but `note_write` accepts `git_oid *` only.

Comment (pks-t) [reply src/notes.c]: Ah, never mind. I've been confused by the naming here. You could've istead just passed in the pointers, but this shall not put your PR on hold now.

File: tests/notes/notes.c [hunk 14]
```diff
@@ -73,6 +73,144 @@ static int note_list_cb(
...
+	}
+
+	cl_fail("Did not see expected note");
+	return 0;
+}
+
+static struct note_create_payload can_create_a_note[] = {
+	{ "1c9b1bc36730582a42d56eeee0dc58673d7ae869", "4a202b346bb0fb0db7eff3cffeb3c70babbd2045", 0 },
+	{ NULL, NULL, 0 }
+};
```
Comment (pks-t) [inline tests/notes/notes.c]: This should be declared inside of the test function itself

File: tests/notes/notes.c [hunk 15]
```diff
@@ -73,6 +73,144 @@ static int note_list_cb(
...
+
+		if (git_oid_cmp(&expected_note_oid, blob_oid) != 0) {
+			continue;
+		}
+
+		cl_git_pass(git_oid_fromstr(&expected_target_oid, notes[i].object_oid));
+
+		if (git_oid_cmp(&expected_target_oid, annotated_obj_id) != 0) {
+			continue;
+		}
```
Comment (pks-t) [inline tests/notes/notes.c]: We usually avoid braces around single-line statements.

File: tests/notes/notes.c [hunk 16]
```diff
@@ -73,6 +73,144 @@ static int note_list_cb(
...
+	}
+
+	cl_assert_equal_i(1, retrieved_notes);
+}
+
+static struct note_create_payload can_create_a_note_from_commit[] = {
+	{ "1c9b1bc36730582a42d56eeee0dc58673d7ae869", "4a202b346bb0fb0db7eff3cffeb3c70babbd2045", 0 },
+	{ NULL, NULL, 0 }
+};
```
Comment (pks-t) [inline tests/notes/notes.c]: Same here

File: tests/notes/notes.c [hunk 17]
```diff
@@ -73,6 +73,123 @@ static int note_list_cb(
...
+	cl_git_pass(git_reference_create(&ref, _repo, "refs/notes/i-can-see-dead-notes", &notes_commit_out, 0, NULL));
+
+	cl_git_pass(git_note_foreach(_repo, "refs/notes/i-can-see-dead-notes", note_list_create_cb, &retrieved_notes));
+
+	cl_assert_equal_i(1, retrieved_notes);
+
+	git_reference_free(ref);
+}
+
+/* Test that we can create a note from a commit, given an existing commit */
```
[the remark this answers is not in the document]
Comment (pks-t) [reply tests/notes/notes.c]: The comment is still there *heh*

File: tests/notes/notes.c [hunk 18]
```diff
@@ -73,6 +73,144 @@ static int note_list_cb(
...
+	 * so we must manually create the ref, that points to the commit */
+	cl_git_pass(git_reference_create(&ref, _repo, "refs/notes/i-can-see-dead-notes", &notes_commit_out, 0, NULL));
+
+	cl_git_pass(git_note_foreach(_repo, "refs/notes/i-can-see-dead-notes", note_list_create_cb, &can_create_a_note_from_commit_given_an_existing_commit));
+
+	for (i = 0; can_create_a_note_from_commit_given_an_existing_commit[i].note_oid != NULL; i++) {
+		if (can_create_a_note_from_commit_given_an_existing_commit[i].seen) {
+			retrieved_notes++;
+		}
+	}
```
Comment (pks-t) [inline tests/notes/notes.c]: Maybe a common function `assert_notes_seens` would be cool here, which simply asserts that all `.seen` fields are set.

File: src/notes.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (pks-t) [reply src/notes.c]: I guess you missed that one on your revisit :)

File: src/notes.c (hunk 10 above)
Comment (pks-t) [reply src/notes.c]: Same here

File: tests/notes/notes.c [hunk 19]
```diff
@@ -431,6 +431,31 @@ void test_notes_notes__can_read_a_note_from_a_commit(void)
...
+	git_commit_free(notes_commit);
+
+	git_commit_lookup(&notes_commit, _repo, &notes_commit_oid);
+
+	cl_assert(notes_commit);
+
+	cl_git_fail_with(GIT_ENOTFOUND, git_note_commit_read(&note, _repo, notes_commit, &oid));
+
+	git_commit_free(notes_commit);
+}
```
Comment (pks-t) [inline tests/notes/notes.c]: :+1:

**Comment by richardipsum** [comment]
@pks-t Thanks for the review! I'm happy to quickly make these changes right now.

File: src/notes.c (hunk 13 above)
Comment (richardipsum) [reply src/notes.c]: Ah I see, so I'm duplicating the `git_oid_cpy`, yes that's a shame, I'll change it.

**Comment by richardipsum** [comment]
@pks-t I've made the requested changes :smile:

**Comment by pks-t** [comment]
Thanks @richardipsum for fixing up! I'm fine with this PR now. I'll wait until the end of the week to see if there are additional comments as this is quite big.

/ping @ethomson, @carlosmn

**Comment by richardipsum** [comment]
@ethomson @carlosmn this was approved by @pks-t a while back, do you guys have thoughts on anything I could do to improve it or could it be merged now? thanks!

**Comment by ethomson** [comment]
I'm :+1: on your commit `9019cdb` which drops the `malloc`.  But unless I'm missing something obvious, where are those OID strings actually used?  Can we drop `target` entirely from those two functions?

**Comment by richardipsum** [comment]
@ethomson Yes you're right, target becomes unused when the commit " Rewrite funcs in terms of note_commit funcs" (94e1dd1) is applied, so should be removed by that commit.

I'll make this change.

**Comment by richardipsum** [comment]
@ethomson Done, I also noticed that some of the whitespace is inconsistent, in some cases I did an extra indent to avoid function arguments lining up with the function name. Not sure what the preferred style for this project is. Happy to modify this again for whitespace consistency if someone tells me what is preferred.

**Comment by ethomson** [comment]
Thanks @richardipsum!  😀

**Comment by joshtriplett** [comment]
@richardipsum Thanks for all your work!

**Comment by pks-t** [comment]
Jay, finally! 🍾 😄
